pkg/recovery/recovery.go

```diff
@@ -0,0 +1,96 @@
+package recovery
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/appscode/go/log"
+	api "github.com/appscode/stash/apis/stash/v1alpha1"
+	cs "github.com/appscode/stash/client/typed/stash/v1alpha1"
+	"github.com/appscode/stash/pkg/cli"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type RecoveryOpt struct {
+	Namespace    string
+	RecoveryName string
+	KubeClient   kubernetes.Interface
+	StashClient  cs.StashV1alpha1Interface
+}
+
+func (opt *RecoveryOpt) RunRecovery() {
+	recovery, err := opt.StashClient.Recoveries(opt.Namespace).Get(opt.RecoveryName, metav1.GetOptions{})
+	if err != nil {
+		log.Infoln(err)
+		return
+	}
+
+	if err = recovery.IsValid(); err != nil {
+		log.Infoln(err)
+		opt.UpdateRecoveryStatus(recovery, "FAILED:"+err.Error())
+		return
+	}
+
+	if err = opt.RecoverOrErr(recovery); err != nil {
+		log.Infoln(err)
+		opt.UpdateRecoveryStatus(recovery, "FAILED:"+err.Error())
+		return
+	}
+
+	opt.UpdateRecoveryStatus(recovery, "SUCCEED")
+}
+
+func (opt *RecoveryOpt) RecoverOrErr(recovery *api.Recovery) error {
+	restic, err := opt.StashClient.Restics(opt.Namespace).Get(recovery.Spec.Restic, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if err = restic.IsValid(); err != nil {
+		return err
+	}
+	if restic.Status.BackupCount < 1 {
+		return fmt.Errorf("no backup found")
+	}
+
+	secret, err := opt.KubeClient.CoreV1().Secrets(opt.Namespace).Get(restic.Spec.Backend.StorageSecretName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	nodeName := os.Getenv("NODE_NAME")
+	if nodeName == "" {
+		log.Fatalln(`Missing ENV var "NODE_NAME"`)
+	}
+	appKind, appName, err := api.ExtractWorkload(recovery.Spec.Workload)
+	if err != nil {
+		return err
+	}
+	podName, _ := api.StatefulSetPodName(appName, recovery.Spec.PodOrdinal) // ignore error for other appKind
+	hostname, smartPrefix, err := api.HostnamePrefixForAppKind(appKind, appName, podName, nodeName)
+	if err != nil {
+		return err
+	}
+
+	cli := cli.New("/tmp", hostname)
+	if err = cli.SetupEnv(restic, secret, smartPrefix); err != nil {
+		return err
+	}
+
+	for _, fg := range restic.Spec.FileGroups {
+		if err = cli.Restore(fg.Path, hostname); err != nil {
```

Comment:
We need to time these calls so that we can set that in the status and also send to Prometheus. 

eg:
https://github.com/appscode/stash/blob/3910849819dc4da098aab7d8aa525ddb76dec2d6/pkg/scheduler/scheduler.go#L287